### ISSUE TITLE
PartDiscovery.CreatePartsAsync() deadlocks when called synchronously from UI thread

### ISSUE THREAD

**Issue by ZombiesWithCoffee** [Microsoft/vs-mef#80]
I've been a fan of MEF for quite a while and very excited to see some development on it.    

My WPF app is .NET Framework 4.6.1.  When using 15.8.55-alpha (I needed the InheritedExportAttribute), the example 'Hosting MEF in an extensible application' just never returns when calling from the App.xaml.cs OnStartup method.

var catalog = ComposableCatalog.Create(Resolver.DefaultInstance)
    .AddParts(discovery.CreatePartsAsync(Assembly.GetExecutingAssembly()).Result)
    .WithCompositionService(); // Makes an ICompositionService export available to MEF parts to import

Can we get a full example of how MEF should be composted in a WPF application?

**Comment by ZombiesWithCoffee** [comment]
Interestingly enough, if I make it async, it does pass thru this code.

                var parts = await discovery.CreatePartsAsync(Assembly.GetExecutingAssembly());

                // Build up a catalog of MEF parts
                var catalog = ComposableCatalog.Create(Resolver.DefaultInstance)
                    .AddParts(parts)
                    .WithCompositionService(); // Makes an ICompositionService export available to MEF parts to import

**Comment by AArnott** [comment]
Hi @ZombiesWithCoffee. I'm pleased to hear it's a useful library for you.

FYI you needn't use 15.8.55-alpha just for `InheritedExportAttribute`. That feature has worked since its initial version so using the latest stable version should be fine.

Regarding the hang you reported, that's because you're invoking an async method then synchronously blocking on its result. That's never a safe technique (although it sometimes works) since the continuations within the async methods you're calling may want to return to the caller's context (which is the default await behavior) and they can't because you're blocking the UI thread. The ideal fix is to be async about it, as you've discovered a way to do already. When you *must* synchronously block for async work to complete, I recommend you use the second of [The 3 threading rules](https://github.com/Microsoft/vs-threading/blob/master/doc/threading_rules.md#rule-2-when-an-implementation-of-an-already-shipped-public-api-must-call-asynchronous-code-and-block-for-its-completion-it-must-do-so-by-following-this-simple-pattern).

Will that work for you?

**Comment by ZombiesWithCoffee** [comment]
I wish it was true about InheritedExport, but I've attempted multiple times via NuGet Package manager and only the very latest alpha version will allow for compilation.  I just tested it 3 times back and forth, and indeed, the release version won't compile.

I believe it has to do with Microsoft.VisualStudio.Composition.NetFxAttributes.  That is included only in the latest alpha release, and I believe it is what is allowing me to compile.  Once I add that (v15.8.55-alpha), then I can use the release version, and the alpha of this one.

As for the startup code and the async, this is how I am currently doing it.  I have to wait for MEF to initialize or the UI comes up before the composition is complete.

        protected override void OnStartup(StartupEventArgs e)
        {
            base.OnStartup(e);

            Task.Run(async () =>{
                await InitializeMef();
            }).Wait();
        }

      async Task<bool> InitializeMef()
        {
            try
            {
                var catalog = await new MefHosting().CreateProductCatalogAsync();

                // Assemble the parts into a valid graph.
                var config = CompositionConfiguration.Create(catalog);

                // Prepare an ExportProvider factory based on this graph.
                var epf = config.CreateExportProviderFactory();

                // Create an export provider, which represents a unique container of values.
                // You can create as many of these as you want, but typically an app needs just one.
                var exportProvider = epf.CreateExportProvider();

                // Obtain our first exported value
                //         var program = exportProvider.GetExportedValue<App>();

**Comment by ZombiesWithCoffee** [comment]
And frankly, at this point I am stumped.  That last 'GetExportedValue<App>' returns a composition error.

I used to call Mef.Build(this) and all of my ViewModels would become instantiated and all of the Imports resolved, but none of my ImportingConstructor classes are being created.  

What magical line of code am I missing?

**Comment by AArnott** [comment]
Yes, moving MEF composition to a background thread before invoking it should allow you to block the UI thread. 

`InheritedExportAttribute` is available in .NET 4.5. While 15.8.55-alpha declares the attribute, it actually just defers to the one in .NET 4.5. If you couldn't use that attribute before, you may have been using the `System.Composition` attributes instead of the `System.ComponentModel.Composition` attributes. Make sure you're using the latter attributes in all cases if you want to use `InheritedExportAttribute`. If you mix attributes between these two namespaces, particularly within the same MEF parts, things will not go the way you might expect.

**Comment by AArnott** [comment]
> I used to call Container.ComposeParts(this) and all of my ViewModels would become instantiated and all of the Imports resolved, but none of my ImportingConstructor classes are being created.

VS-MEF doesn't have a `ComposeParts` method. But it does support `SatisfyImportsOnce`, which may solve your requirement. [Read more on this](https://github.com/Microsoft/vs-mef/blob/master/doc/dynamic_recomposition.md)

**Comment by ZombiesWithCoffee** [comment]
I'm uninstalling all of this.  Thanks for the help AArnott!   I'll come back when there is more documentation.  11 hours is too much to spend on it.

**Comment by AArnott** [comment]
Sorry to hear of how long you spent.

**Comment by AArnott** [comment]
I'm going to re-open this to track adding the `AsyncUsageAnalyzers` to the project and considering violations of the `UseConfigureAwait` rule to be an error.

**Comment by ZombiesWithCoffee** [comment]
Good idea.  I'm positive you know, but the code used was just from this doc page.

https://github.com/Microsoft/vs-mef/blob/master/doc/hosting.md

**Comment by ZombiesWithCoffee** [comment]
I did try the AppHost package as well.  It did create the files in the obj/bin directory, but always crashed as soon as it ran.  I'm sorry I didn't write down the error messages.
